KAFKA-20335: Document KIP-1271 headers-aware state stores (#21840)
apache/kafka: docs/streams/developer-guide/interactive-queries.md

Reviewer: > Kafka Streams currently has two built-in types:

This need to get updated. "two" was outdated even before we added header stores. It's eight now with header stores.

Author: updated with all 8 here.

Reviewer: Seems this part is also outdated. `This example creates a key-value store` -- as it's a DSL example, it actually creates a timestamped store.

Reviewer: Similar further below: `This example creates a window store`

Reviewer: Guess we can still keep the code snippets as-is and query they stores for "plain values" only, but we might want to add a note about it.

Not supported (you must configure)
</td> </tr> </table>

# Querying local state stores for an app instance {#querying-local-state-stores-for-an-app-instance}

A Kafka Streams application typically runs on multiple instances. The state that is locally available on any given instance is only a subset of the [application's entire state](../architecture.html#streams-architecture-state). Querying the local stores on an instance will only return data locally available on that particular instance.

The method `KafkaStreams#store(...)` finds an application instance's local state stores by name and type. Note that interactive queries are not supported for [versioned state stores](processor-api.html#streams-developer-guide-state-store-versioned) at this time.

> Headers-aware state stores added in Kafka 4.3 ([KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)) work with the same **`QueryableStoreTypes`** helpers as headerless stores—for example `timestampedKeyValueStore()` and `timestampedWindowStore()` also match the corresponding `*WithHeaders` implementations, and `sessionStore()` applies to `SessionStoreWithHeaders`. However, **`KafkaStreams#store(...)`** still exposes the usual read-only shapes: plain values or **`ValueAndTimestamp`** for timestamped key-value and window stores; **record headers are not included** in those IQv1 results (facades strip them). Session queries return only the aggregation value, not header metadata. **[`ValueTimestampHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/ValueTimestampHeaders.html)** is what processors see when reading the store inside the topology. IQv2 may still omit headers in some query paths—see the KIP, release notes, and [Processor API: Headers in State Stores](processor-api.html#headers-in-state-stores).

Reviewer: > helpers as headerless stores—for example 

Not sure if the dash (which I believe is supposed to be an m-dash will render correctly? Can we confirm?

> However, **`KafkaStreams#store(...)`** still exposes the usual read-only shapes

Well, depends what queryable-store-type you pick. You can also get the headers if you want. -- In combination with the later `**[`ValueTimestampHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/ValueTimestampHeaders.html)** is what processors see when reading the store inside the topology.` it seems you say that you cannot get headers at all with IQv1, but you can.

> plain values or **`ValueAndTimestamp`** for timestamped key-value and window stores

-> `for (timestamped) key-value`

> (facades strip them)

I would remove this -- implementation detail.

> IQv2 may still omit headers in some query paths—see the KIP, release notes, and [Processor API: Headers in State Stores](processor-api.html#headers-in-state-stores).

IQv2 is not implemented with AK 4.3 yet -- and we will need follow up KIPs to add support.

Maybe the whole paragraph can we simplified a little bit? It add good context that "richer" stores (ts, header) can we queried with "lower" queryable store types, but that's not header specific.

Author: verified m-das works. Updated this part for IQV1 and removed any IQV2 reference here.

Not supported (you must configure)
</td> </tr> </table>

# Querying local state stores for an app instance {#querying-local-state-stores-for-an-app-instance}

A Kafka Streams application typically runs on multiple instances. The state that is locally available on any given instance is only a subset of the [application's entire state](../architecture.html#streams-architecture-state). Querying the local stores on an instance will only return data locally available on that particular instance.

The method `KafkaStreams#store(...)` finds an application instance's local state stores by name and type. Note that interactive queries are not supported for [versioned state stores](processor-api.html#streams-developer-guide-state-store-versioned) at this time.

> Headers-aware state stores added in Kafka 4.3 ([KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)) use the same `QueryableStoreTypes` helpers as the corresponding stores without headers. For example, `timestampedKeyValueStore()` and `timestampedWindowStore()` also match `TimestampedKeyValueStoreWithHeaders` and `TimestampedWindowStoreWithHeaders`, and `sessionStore()` matches `SessionStoreWithHeaders`.

Reviewer 2: We are talking about `QueryableStoreTypes`, but they are introduced later in the doc.

> Headers-aware state stores added in Kafka 4.3 ([KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)) use the same `QueryableStoreTypes` helpers as the corresponding stores without headers. For example, `timestampedKeyValueStore()` and `timestampedWindowStore()` also match `TimestampedKeyValueStoreWithHeaders` and `TimestampedWindowStoreWithHeaders`, and `sessionStore()` matches `SessionStoreWithHeaders`.

Reviewer 2: mixed terminology: "Headers-aware" vs. "stores without headers".  Pick one please.

The result returned by `KafkaStreams#store(...)` depends on the `QueryableStoreType`. The built-in `QueryableStoreTypes` facades (`keyValueStore()`, `timestampedKeyValueStore()`, `windowStore()`, `timestampedWindowStore()`, `sessionStore()`) expose read-only views that return values, or `ValueAndTimestamp` for timestamped key-value and window stores. Record headers are not included in these results. A custom `QueryableStoreType` can expose the underlying header-aware store; see [KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores).

> 

Reviewer 2: Calling the static factories "facades" is technically off — they're factory methods that return QueryableStoreType matchers. Suggest "factory methods" or just "helpers". Also the sentence is parenthetical-heavy; split it.

> Within the topology, processors can access headers through types such as [`ValueTimestampHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/ValueTimestampHeaders.html) for key-value and window stores, and [`AggregationWithHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/AggregationWithHeaders.html) for session stores; see [Headers in State Stores](processor-api.html#headers-in-state-stores).

Reviewer 2: This is about Processor-API-side access, not interactive queries. It's a useful cross-reference, but it interrupts the IQ flow. A short trailing sentence with a link would be better than a full paragraph.

> Within the topology, processors can access headers through types such as [`ValueTimestampHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/ValueTimestampHeaders.html) for key-value and window stores, and [`AggregationWithHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/AggregationWithHeaders.html) for session stores; see [Headers in State Stores](processor-api.html#headers-in-state-stores).

Reviewer 2: `Within the topology, processors can access headers through types such as …` — the processor doesn't access headers directly; the store returns objects of those types when accessed from the processor. Suggestion: "Inside processor code, header-aware stores expose record headers via …".

> 

Reviewer 2: The 3 paragraphs discuss QueryableStoreTypes helpers before the doc has introduced what those helpers are. They wedge themselves between the KafkaStreams#store(...) paragraph and the architecture image at line 138. The image originally illustrated the previous paragraph; now it's separated by ~3 dense paragraphs and feels orphaned.

![](/43/images/streams-interactive-queries-api-01.png)

Every application instance can directly query any of its local state stores.

The _name_ of a state store is defined when you create the store. You can create the store explicitly by using the Processor API or implicitly by using stateful operations in the DSL.

The _type_ of a state store is defined by `QueryableStoreType`. You can access the built-in store types via [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html). The supported types are:

>   * **`QueryableStoreTypes#keyValueStore()`** — [`KeyValueStoreType`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.KeyValueStoreType.html). See [Querying local key-value stores](#querying-local-key-value-stores).

Reviewer 2: The pattern method()  — TypeLink. See sectionLink. reads as three loosely-related items. The relationship between the method and the type isn't explained (return type? matcher class?). The original two-bullet form was more readable; consider keeping the prose tight: `keyValueStore()` — see [Querying local key-value stores] and only link to the type if it adds value.

  * **`QueryableStoreTypes#timestampedKeyValueStore()`** — [`TimestampedKeyValueStoreType`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.TimestampedKeyValueStoreType.html). See [Querying local key-value stores](#querying-local-key-value-stores).
  * **`QueryableStoreTypes#windowStore()`** — [`WindowStoreType`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.WindowStoreType.html). See [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#timestampedWindowStore()`** — [`TimestampedWindowStoreType`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.TimestampedWindowStoreType.html). See [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#sessionStore()`** — [`SessionStoreType`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.SessionStoreType.html). See [Querying local window stores](#querying-local-window-stores).

> Header-aware store variants (introduced in [KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)) are queried using the same matchers as their corresponding timestamped or session stores.

Reviewer 2: ```suggestion
Headers-aware store variants (introduced in [KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)) are queried using the same matchers as their corresponding timestamped or session stores.
```

> 

Reviewer 2: This is a near-duplicate of line 132. One of them should be removed.

Author: changed it.

The _name_ of a state store is defined when you create the store. You can create the store explicitly by using the Processor API or implicitly by using stateful operations in the DSL.

The _type_ of a state store is defined by `QueryableStoreType`. Pass a built-in implementation from [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html) as the second argument to `KafkaStreams#store(...)`. The helpers below are those implementations:

  * **`QueryableStoreTypes#keyValueStore()`** — see [Querying local key-value stores](#querying-local-key-value-stores).

>   * **`QueryableStoreTypes#timestampedKeyValueStore()`** — see [Querying local key-value stores](#querying-local-key-value-stores).

Reviewer: We should also list `QueryableStoreTypes#timestampedKeyValueStoreWithHeaders()` here?

I don't see why we would want to make the headers case "special"? -- Same for window and session case.

Author: added.

  * **`QueryableStoreTypes#windowStore()`** — see [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#timestampedWindowStore()`** — see [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#sessionStore()`** — see [Querying local window stores](#querying-local-window-stores).

> ### Header-aware stores and interactive queries (Kafka 4.3+) {#header-aware-stores-interactive-queries}

Reviewer: ```suggestion
### Header-aware stores and interactive queries {#header-aware-stores-interactive-queries}
```

Reviewer: This change only goes into docs for AK 4.3, so no need to call out the version.

> Since Kafka 4.3, you can use [header-aware state stores](processor-api.html#headers-in-state-stores) ([KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)). For interactive queries, they reuse the same [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html) helpers as matching **headerless** stores. `timestampedKeyValueStore()` and `timestampedWindowStore()` also match `TimestampedKeyValueStoreWithHeaders` and `TimestampedWindowStoreWithHeaders`. `sessionStore()` matches `SessionStoreWithHeaders`.

Reviewer: I would remove the KIP reference. It's not the upgrade guide. If we add KIP reference for every feature across the docs, it becomes a mess.

Author: removed.

> Since Kafka 4.3, you can use [header-aware state stores](processor-api.html#headers-in-state-stores) ([KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)). For interactive queries, they reuse the same [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html) helpers as matching **headerless** stores. `timestampedKeyValueStore()` and `timestampedWindowStore()` also match `TimestampedKeyValueStoreWithHeaders` and `TimestampedWindowStoreWithHeaders`. `sessionStore()` matches `SessionStoreWithHeaders`.

Reviewer: > Since Kafka 4.3

Similar. No need to have version references "everywhere".

Reviewer: What the paragraph says is correct, but not really meaningful? In the end, we add new "withheader" variations for `QueryableStoreTypes` and of course the existing one can be used, too. But that is nothing "new". You can also use `QueryableStoreTypes#keyValueStore()` agains a ts-ks-store.

Author: removed.

> The return type of `KafkaStreams#store(...)` follows the `QueryableStoreType` you pass in. Built-in helpers yield read-only facades with plain values, or `ValueAndTimestamp` for timestamped key-value and window stores. They do not expose record headers. To include headers in IQ results, implement a custom `QueryableStoreType` (see [KIP-1271](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1271%3A+Allow+to+Store+Record+Headers+in+State+Stores)). For [`ValueTimestampHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/ValueTimestampHeaders.html) and [`AggregationWithHeaders`](/{version}/javadoc/org/apache/kafka/streams/state/AggregationWithHeaders.html) in the Processor API, see [Headers in State Stores](processor-api.html#headers-in-state-stores).

Reviewer: > To include headers in IQ results, implement a custom `QueryableStoreType`

That's not necessary. Cf my other comment.

We can also remove the KIP reference.

The _name_ of a state store is defined when you create the store. You can create the store explicitly by using the Processor API or implicitly by using stateful operations in the DSL.

> The _type_ of a state store is defined by `QueryableStoreType`. Pass a built-in implementation from [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html) as the second argument to `KafkaStreams#store(...)`. The helpers below are those implementations:

Reviewer 2: ```suggestion
The _type_ of a state store is defined by `QueryableStoreType`. Pass a built-in implementation from [`QueryableStoreTypes`](/{version}/javadoc/org/apache/kafka/streams/state/QueryableStoreTypes.html) as the second argument to `KafkaStreams#store(...)`. The available built-in helpers are:
```

  * **`QueryableStoreTypes#keyValueStore()`** — see [Querying local key-value stores](#querying-local-key-value-stores).
  * **`QueryableStoreTypes#timestampedKeyValueStore()`** — see [Querying local key-value stores](#querying-local-key-value-stores).
  * **`QueryableStoreTypes#timestampedKeyValueStoreWithHeaders()`** — see [#header-aware-stores-interactive-queries](#header-aware-stores-interactive-queries).
  * **`QueryableStoreTypes#windowStore()`** — see [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#timestampedWindowStore()`** — see [Querying local window stores](#querying-local-window-stores).
  * **`QueryableStoreTypes#timestampedWindowStoreWithHeaders()`** — see [#header-aware-stores-interactive-queries](#header-aware-stores-interactive-queries).
  * **`QueryableStoreTypes#sessionStore()`** — see [Querying local window stores](#querying-local-window-stores).

>   * **`QueryableStoreTypes#sessionStoreWithHeaders()`** — see [#header-aware-stores-interactive-queries](#header-aware-stores-interactive-queries).

Reviewer 2: In lines 142, 145, and 147:  The link text is the raw anchor. Replace with `see [Header-aware stores and 
  interactive queries](#header-aware-stores-interactive-queries)` to match the target heading defined at line 149.

Author: Replaced with see [Header-aware stores and  interactive queries](#header-aware-stores-interactive-queries)

### Header-aware stores and interactive queries {#header-aware-stores-interactive-queries}

> For a [header-aware store](processor-api.html#headers-in-state-stores), use the matching **`*WithHeaders()`** entry from the list above when interactive query results must include headers.

Reviewer 2: ```suggestion
For a [header-aware store](processor-api.html#headers-in-state-stores), use the **`*WithHeaders()`** entry that corresponds to your store type when query results must include headers.
```

Reviewer 2: `matching` is a bit vague.

Reviewer 2: @nileshkumar3 what about this comment?

Reviewer 2: Do you please remove the `matching` everywhere and replcae it with a better sentence/term.

> ## Querying local key-value stores

Reviewer 2: Why is that in h2? I see  heading-level jump (h1 (line 126) → h3 (line 149)→ h2: this line)

Reviewer 2: The document goes h1 (line 126) → h3 (line 149) → h2 (line 161). Skipping a level (no h2 between the h1 and the h3) breaks the heading hierarchy. Either
- Promote to h2 so it sits as a sibling of ## Querying local key-value stores, ## Querying local window stores, ## Querying local custom state stores. The whole "Querying local …" family
  lines up under the h1.
- Demote to a **Note:** block (no heading at all) appended to the bullet list, since the body is just one sentence. This is what I'd actually recommend — a heading for one sentence is
  overkill.

Reviewer 2: Another issue: The new ### Header-aware stores… heading at line 149 splits the intro in half. The two general lines that follow it — "You can also implement your own QueryableStoreType…" and the Note about per-partition stores — apply to all QueryableStoreTypes, but the new heading makes them visually belong to the header-aware subsection only. Move those two lines above the new heading (or drop the heading entirely) so they stay part of the general intro.

Author: Updated: dropped the extra heading, kept the anchor + Note, general intro stays above it.

## Querying local key-value stores

To query key-value state, you first build a topology that includes a state store. This example uses the DSL `count()` operator on a grouped stream, which creates a timestamped key-value store named `CountsKeyValueStore`. That store holds the latest count for each word from the topic `word-count-input`.

> Note: The snippets below still use `QueryableStoreTypes.keyValueStore()` and `ReadOnlyKeyValueStore<String, Long>`, which return values only (the counts). The state store keeps values together with their timestamps; use `QueryableStoreTypes.timestampedKeyValueStore()` and `ReadOnlyKeyValueStore<String, ValueAndTimestamp<Long>>` if you need access to timestamps via interactive queries.

Reviewer 2: `The snippets below still use..`?  why `still`? Did we introduce it back then in the doc? this is the first one in the section
